File: packages/guarddog-ios-sdk/GuardDogCore/Sources/GuardDogCore/Net/UrlSanitizer.swift
import Foundation

/// URL sanitization (Swift). Parity: security/test-vectors/normalization/url_vectors.json
/// The original candidate stays analyzable; `sanitizedUrl` = scheme://canonicalHost + path only.
public struct ParsedCandidateUrl: Equatable {
    public let original: String
    public let scheme: String
    public let host: String
    public let path: String
    public let hadUserinfo: Bool
    public let hadQuery: Bool
    public let hadFragment: Bool
    public var sanitizedUrl: String { "\(scheme)://\(host)\(path)" }
}

public enum UrlSanitizer {
    static let urlRegex = try! NSRegularExpression(pattern: "^([a-zA-Z][a-zA-Z0-9+.-]*)://([^/?#]*)([^?#]*)(\\?[^#]*)?(#.*)?$")
    static let digitsRegex = try! NSRegularExpression(pattern: "^\\d*$")

    public static func sanitize(_ raw: String) -> ParsedCandidateUrl? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        let ns = trimmed as NSString
        guard let m = urlRegex.firstMatch(in: trimmed, range: NSRange(location: 0, length: ns.length)) else { return nil }
        func group(_ i: Int) -> String { let r = m.range(at: i); return r.location == NSNotFound ? "" : ns.substring(with: r) }
        let scheme = group(1).lowercased()
        guard scheme == "http" || scheme == "https" else { return nil }
        var authority = group(2)
        let hadUserinfo = authority.contains("@")
        if hadUserinfo, let at = authority.range(of: "@", options: .backwards) { authority = String(authority[at.upperBound...]) }
        let hostPart: String
        if authority.hasPrefix("[") {
            guard let end = authority.firstIndex(of: "]") else { return nil }
            hostPart = String(authority[...end])
            let rest = String(authority[authority.index(after: end)...])
            if !rest.isEmpty && !(rest.hasPrefix(":") && HostCanonicalizer.matches(digitsRegex, String(rest.dropFirst()))) { return nil }
        } else if let colon = authority.range(of: ":", options: .backwards) {
            hostPart = String(authority[..<colon.lowerBound])
            if !HostCanonicalizer.matches(digitsRegex, String(authority[colon.upperBound...])) { return nil }
        } else {
            hostPart = authority
        }
        guard let host = HostCanonicalizer.canonicalize(hostPart) else { return nil }
        let path = group(3).isEmpty ? "/" : group(3)
        return ParsedCandidateUrl(original: raw, scheme: scheme, host: host, path: path, hadUserinfo: hadUserinfo,
                                  hadQuery: !group(4).isEmpty, hadFragment: !group(5).isEmpty)
    }
}

File: packages/guarddog-android-sdk/guarddog-core/src/main/java/com/guarddog/core/net/UrlSanitizer.kt
package com.guarddog.core.net

/**
 * URL sanitization (Kotlin). Parity: security/test-vectors/normalization/url_vectors.json
 *
 * The ORIGINAL candidate stays available for local analysis. The share-safe form is
 * `scheme://canonicalHost + path` - no userinfo, no port, no query, no fragment.
 * Only `sanitizedUrl` may appear in SecurityEvent or leave the device.
 */
data class ParsedCandidateUrl(
    val original: String,
    val scheme: String,
    val host: String,
    val path: String,
    val hadUserinfo: Boolean,
    val hadQuery: Boolean,
    val hadFragment: Boolean,
) {
    val sanitizedUrl: String get() = "$scheme://$host$path"
}

object UrlSanitizer {
    private val urlRegex = Regex("^([a-zA-Z][a-zA-Z0-9+.-]*)://([^/?#]*)([^?#]*)(\\?[^#]*)?(#.*)?$")
    private val digits = Regex("^\\d*$")

    fun sanitize(raw: String): ParsedCandidateUrl? {
        val m = urlRegex.matchEntire(raw.trim()) ?: return null
        val scheme = m.groupValues[1].lowercase()
        if (scheme != "http" && scheme != "https") return null
        var authority = m.groupValues[2]
        val hadUserinfo = authority.contains('@')
        if (hadUserinfo) authority = authority.substring(authority.lastIndexOf('@') + 1)
        val hostPart: String
        if (authority.startsWith("[")) {
            val end = authority.indexOf(']')
            if (end < 0) return null
            hostPart = authority.substring(0, end + 1)
            val rest = authority.substring(end + 1)
            if (rest.isNotEmpty() && !(rest.startsWith(":") && digits.matches(rest.substring(1)))) return null
        } else {
            val colon = authority.lastIndexOf(':')
            hostPart = if (colon >= 0) authority.substring(0, colon) else authority
            if (colon >= 0 && !digits.matches(authority.substring(colon + 1))) return null
        }
        val host = HostCanonicalizer.canonicalize(hostPart) ?: return null
        val path = m.groupValues[3].ifEmpty { "/" }
        return ParsedCandidateUrl(
            original = raw,
            scheme = scheme,
            host = host,
            path = path,
            hadUserinfo = hadUserinfo,
            hadQuery = m.groupValues[4].isNotEmpty(),
            hadFragment = m.groupValues[5].isNotEmpty(),
        )
    }
}
